### PR TITLE
Master command palette dialog2 fge

### PR DESCRIPTION
**Pull request by FrancoisGe**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by aab-odoo — APPROVED** [review]
Maybe add some tests?

File: addons/web/static/src/core/utils/search.js
Comment (aab-odoo) [inline addons/web/static/src/core/utils/search.js:55]: should not sort imo

**Review by aab-odoo — APPROVED** [review]

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (seb-odoo) [inline addons/mail/static/src/webclient/commands/mail_providers.js:18]: guideline: closing tick of the xml preferably on the same indent level as its opening tick

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (seb-odoo) [inline addons/mail/static/src/webclient/commands/mail_providers.js:16]: guideline: unnecessary space before `/>`

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (seb-odoo) [inline addons/mail/static/src/webclient/commands/mail_providers.js:13]: Isn't this `<t>` tag useless when giving the template directly?
`t-name` shouldn't be necessary, and `owl="1"` is only used during templates loading from assets.

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (seb-odoo) [inline addons/mail/static/src/webclient/commands/mail_providers.js:14]: `o_command_default`? Isn't this specifically not a default one?

Also I don't understand what `o_command_dialog` is supposed to mean, and it doesn't look used anywhere.

Edit: ok I found it and I realized the rule is build dynamically with string manipulation, this is really not ideal for ease of understanding.

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (seb-odoo) [inline addons/mail/static/src/webclient/commands/mail_providers.js:38]: Clients usually open tickets when we sort on strings without ignoring case and accents, so I would suggest doing it (or more ideally just following whatever order the server returned and giving the order param to the rpc if necessary).

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (seb-odoo) [inline addons/mail/static/src/webclient/commands/mail_providers.js:33]: It's strange to find `name` and `email` at the same level as the `Component` and `action` definition, I would have expected to find them on a `props` key maybe?

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (seb-odoo) [inline addons/mail/static/src/webclient/commands/mail_providers.js:27]: Just a note: mixed usage of `await` and `callback` looks really strange here, but as we discussed it's ok to do the clean up of `imSearch` separately.

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (seb-odoo) [reply addons/mail/static/src/webclient/commands/mail_providers.js:52]: `threads` can be named specifically `channels` in this context

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (seb-odoo) [inline addons/mail/static/src/webclient/commands/mail_providers.js:54]: no need for double arrays, the method could directly return a map() on threads

**Review by seb-odoo — CHANGES_REQUESTED** [review]
Thank you for the PR, this will be very useful in practice.

- I get `Uncaught Promise > provide is not a function` when selecting a user, we really need to add at least basic tests before merging it.
- I tested the rpc on a decently sized db and it looked good enough so ok for the current approach.
- Side note: when empty it says "No action found". I would be expecting it to say "No users found" or "No channels found" according to the name space maybe?
- See below for code clean up and other comments.

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (FrancoisGe) [reply addons/mail/static/src/webclient/commands/mail_providers.js:38]: Do you want to modify the SQL query to order the data?

File: addons/web/static/src/webclient/commands/command_palette.js
Comment (brboi) [inline addons/web/static/src/webclient/commands/command_palette.js:194]: await missing

File: addons/web/static/src/webclient/commands/command_palette.xml
Comment (brboi) [inline addons/web/static/src/webclient/commands/command_palette.xml:22]: est-ce qu'on ne voudrait pas plutôt toujours passer le name:
```suggestion
                  <t t-component="command.Component || DefaultCommandItem" label="command.name" t-props="command.props" t-on-close="props.closeMe()" t-on-execute-command="executeCommand(command)"/>
```

ou encore le passer mais en default slot:
```suggestion
                  <t t-component="command.Component || DefaultCommandItem" t-props="command.props" t-on-close="props.closeMe()" t-on-execute-command="executeCommand(command)">
                      <t t-esc="command.name"/>
                  </t>
```

Et du coup les props seront toujours facultatives ?
```ts
interface Command {
    name: string;
    action: Function;
    category?: string;
    Component?: Component;
    props?: Object;
}
```

**Review by seb-odoo — COMMENTED** [review]
Few more code clean up comments but overall it's looking good.

I still have to make a final test after this.

File: addons/mail/static/src/models/discuss/discuss.js
Comment (seb-odoo) [inline addons/mail/static/src/models/discuss/discuss.js:81]:
```suggestion
            const threads = await this.env.models['mail.thread'].searchChannelsToOpen({ keyword: req.term, limit: 10 });
```

File: addons/mail/static/src/models/discuss/discuss.js
Comment (seb-odoo) [inline addons/mail/static/src/models/discuss/discuss.js:88]: `id` is missing from here, selecting a channel that is already joined in discuss sidebar leads to a crash

Comment (seb-odoo) [reply addons/mail/static/src/models/discuss/discuss.js:88]: I'm thinking apart from `label` and `value` which are used by the lib, `id` might be the only necessary field, not `channel_type` and `name` (if referring to usage of `ui.item`).

File: addons/mail/static/src/models/discuss/discuss.js
Comment (seb-odoo) [inline addons/mail/static/src/models/discuss/discuss.js:83]: This line must be kept, the name of a new channel is still empty after creating it in this flux.

File: addons/mail/static/src/webclient/commands/mail_providers.js
Comment (seb-odoo) [inline addons/mail/static/src/webclient/commands/mail_providers.js:24]: no need for extra new line

File: addons/mail/static/tests/helpers/mock_server.js
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:1733]: I don't think that works, one `sort` will probably not be stable compared to the other.
Both conditions should be part of only one sort method: if name is equeal, return the condition on id.
Note that it would be better to compare ids instead of objects (I have no idea what difference between objects like you did it does).

**Review by seb-odoo — CHANGES_REQUESTED** [review]
Still a problem with the change in discuss (I guess we really need to add a test for the problematic case on our side), but otherwise it should be good to merge tomorrow if the last comments are applied.

**Comment by seb-odoo** [comment]
"in front of the username, display the IM status" this part of the task is not done? Because you can't use our im status component with the new env?

**Comment by FrancoisGe** [comment]
> "in front of the username, display the IM status" this part of the task is not done? Because you can't use our im status component with the new env?

Yes, Nathan is ok with that. We can do it in another task.

File: addons/mail/static/tests/helpers/mock_server.js
Comment (FrancoisGe) [reply addons/mail/static/tests/helpers/mock_server.js:1733]: My bad, I want to do 
```js
.sort((a, b) => (a.id - b.id))
.sort((a, b) => (a.name > b.name) ? 1 : -1);
```

File: addons/mail/static/src/models/discuss/discuss.js
Comment (FrancoisGe) [reply addons/mail/static/src/models/discuss/discuss.js:83]: I move it in "searchChannelsToOpen". It's ok ?

File: addons/mail/static/src/models/discuss/discuss.js
Comment (FrancoisGe) [reply addons/mail/static/src/models/discuss/discuss.js:88]: I just adapt the code to have the same object.

Comment (FrancoisGe) [reply addons/mail/static/src/models/discuss/discuss.js:88]: I will remove `channel_type` and `name`

File: addons/web/static/src/webclient/commands/command_palette.js
Comment (brboi) [inline addons/web/static/src/webclient/commands/command_palette.js:22]: action may return a provider and placeholder

Comment (brboi) [inline addons/web/static/src/webclient/commands/command_palette.js:22]: action may return a provide fn with or without a placeholder

File: addons/web/static/src/webclient/commands/command_palette.js
Comment (brboi) [inline addons/web/static/src/webclient/commands/command_palette.js:24]:
```suggestion
 *  Component?: owl.Component,
 *  props?: Object,
```

Comment (brboi) [reply addons/web/static/src/webclient/commands/command_palette.js:24]: As `Component` is optional, `props` should be optional too.

Thus the need, IMHO, to pass `name` outside of the `props` scope...

And then pass `commandItem.name` in the command palette template either as a prop:
```xml
<t t-component="commandItem.Component" name="commandItem.name" t-props="commandItem.props"/>
```
... either as the default slot content:
```xml
<t t-component="commandItem.Component" t-props="commandItem.props">
    <t t-esc="commandItem.name"/>
</t>
```

I let you decide on that, but there is my opinion :-)

File: addons/web/static/src/webclient/commands/command_service.js
Comment (brboi) [inline addons/web/static/src/webclient/commands/command_service.js:118]:
```suggestion
             * @param {HTMLElement} activeElement
```

**Review by brboi — COMMENTED** [review]
I only looked at the command palette, and for me it is OK.
Just apply small comments and decide on the big one :-)

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:594]: I would prefer `searchTerm` instead of `keyword` as name.

File: addons/mail/static/tests/helpers/mock_server.js
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:1732]: lint: missing ;

File: addons/mail/static/tests/webclients/commands/mail_providers_tests.js
Comment (seb-odoo) [inline addons/mail/static/tests/webclients/commands/mail_providers_tests.js:5]: lint: not used

File: addons/mail/static/tests/webclients/commands/mail_providers_tests.js
Comment (seb-odoo) [inline addons/mail/static/tests/webclients/commands/mail_providers_tests.js:66]: lint: no space after first `(`, no need for `{}` on the arrow function

File: addons/mail/static/tests/webclients/commands/mail_providers_tests.js
Comment (seb-odoo) [inline addons/mail/static/tests/webclients/commands/mail_providers_tests.js:68]: lint: missing ;

File: addons/mail/static/tests/webclients/commands/mail_providers_tests.js
Comment (seb-odoo) [inline addons/mail/static/tests/webclients/commands/mail_providers_tests.js:107]: same remarks as above here

**Review by seb-odoo — CHANGES_REQUESTED** [review]
Good to go functionally, including from Discuss app.

Last code check below, mostly lint issue in test file.

I'm assuming the changes in `web` are approved on your side, I see nothing too wrong but I don't know the code as well.

File: addons/web/static/src/webclient/commands/command_palette.js
Comment (brboi) [inline addons/web/static/src/webclient/commands/command_palette.js:25]: missing Component key :-)

File: addons/web/static/src/webclient/commands/command_palette.js
Comment (brboi) [inline addons/web/static/src/webclient/commands/command_palette.js:22]: This should be written like that in order to work properly
```suggestion
 *  action: ()=>(void | NextProvider),
```

File: addons/web/static/src/webclient/commands/command_palette.xml
Comment (brboi) [inline addons/web/static/src/webclient/commands/command_palette.xml:44]: wrong template naming convention ?

I think it should be:
`module_name_in_snake_case.TemplatePartInPascalCase`

**Review by seb-odoo — APPROVED** [review]

**Review by seb-odoo — APPROVED** [review]
@robodoo r+ rebase-ff

**Pull request closed without merging**
